### ISSUE TITLE
Loading kubectl plugin is slow

### ISSUE THREAD

**Issue by lbolla** [ohmyzsh/ohmyzsh#6843]
The main reason is that it regenerates the completion script over and over, with `kubectl completion zsh`.

**Comment by alirezaDavid** [comment]
I'm using kubectl plugin and every morning start terminal it's been hang up about 5 minute.
After remove the plugin from `.zshrc` everything working smoothly.

**Comment by gowtham-sai** [comment]
@lbolla I don't think, it is completely fixed or something else is wrong with mysetup? 
In my machine, it still takes lot of time to load. Here is the profiling screenshot. 

<img width="1680" alt="screen shot 2018-08-20 at 8 02 21 am" src="https://user-images.githubusercontent.com/8710113/44317562-a6a4f400-a44f-11e8-86ff-a9502ad5e040.png">

**Comment by gowtham-sai** [comment]
Avoid it by doing lazy load. 
```bash
function kubectl() {
    if ! type __start_kubectl >/dev/null 2>&1; then
        source <(command kubectl completion zsh)
    fi

    command kubectl "$@"
}
```
as mentioned [here](https://github.com/kubernetes/kubernetes/issues/59078#issuecomment-363384825)

**Comment by lbolla** [comment]
Do you have custom zsh cache settings? kubectl speedup works by caching a file in $ZSH_CACHE_DIR and skipping `kubectl completion zsh` which is the slowest bit.

**Comment by gowtham-sai** [comment]
@lbolla I don't have any custom zsh cache settings. yes I do have the cached file in $ZSH_CACHE_DIR

**Comment by mati865** [comment]
Experienced the same on Fedora, the cache doesn't help because the slow this here is executing function`__kubectl_bash_source()` from `${ZSH_CACHE_DIR}/kubectl_completion`.

Upstream issue: https://github.com/kubernetes/kubernetes/issues/59078

**Comment by akaihola** [comment]
...and if I understand correctly, the lazy loading trick doesn't help either because sourcing the cached `kubectl_completion` triggers loading anyway.

**Comment by wmcnamee-tunein** [comment]
any news on improvements here?

```
num  calls                time                       self            name
-----------------------------------------------------------------------------------
 1)    1         190.03   190.03   65.52%    189.83   189.83   65.45%  __kubectl_bash_source
```

**Comment by clebs** [comment]
I am having the same problem in Mac OSX Catalina 10.15.7 since yesterday when I updated via `omz update`.

The debug startup freezes for some seconds in the following lines:
```
+/Users/xxxx/.oh-my-zsh/cache/kubectl_completion:12832> __kubectl_convert_bash_to_zsh
...
+/Users/xxxx/.oh-my-zsh/cache/minicube_completion:2673> __minikube_convert_bash_to_zsh
```

**Comment by robbyrussell** [comment]
Anyone interested in submitting a PR for this?

**Comment by esemeniuc** [comment]
Running the speed test from https://blog.mattclemente.com/2020/06/26/oh-my-zsh-slow-to-load/#how-to-test-your-shell-load-time,
I see the following (Arch linux, kernel 5.14, zsh 5.8, alacritty terminal, nvme ssd):

```shell
➜  ~ for plugin ($plugins); do
  timer=$(($(date +%s%N)/1000000))
  if [ -f $ZSH_CUSTOM/plugins/$plugin/$plugin.plugin.zsh ]; then
    source $ZSH_CUSTOM/plugins/$plugin/$plugin.plugin.zsh
  elif [ -f $ZSH/plugins/$plugin/$plugin.plugin.zsh ]; then
    source $ZSH/plugins/$plugin/$plugin.plugin.zsh
  fi
  now=$(($(date +%s%N)/1000000))
  elapsed=$(($now-$timer))
  echo $elapsed ms":" $plugin
done

22 ms: git
17 ms: fzf
7 ms: tmux
239 ms: kubectl
2 ms: docker
```

**Comment by marckhouzam** [comment]
I tried this on my Mac (using `gdate` instead of `date`) and I also got over 200 ms:
```
220 ms: kubectl
```
After investigating I noticed that the zsh plugin for kubectl uses a cache:
```
$ ls -l ${ZSH_CACHE_DIR}/kubectl_completion
4 Nov  2019 /Users/marckhouzam/.oh-my-zsh/cache/kubectl_completion
```
which was two years old and using the old kubectl completion logic.
After removing that file, I re-ran the timing script twice (the first time caused the plugin to regenerate the file so it was slower), and on the second and all future measurements I got something like
```
13 ms: kubectl
```
I took this opportunity to empty that cache for other plugins 😄

**Comment by marckhouzam** [comment]
Oh and you need to use kubectl 1.22 or higher for the faster zsh completion logic.

**Comment by esemeniuc** [comment]
Thank you very much @marckhouzam!! It works :partying_face: 

```shell
➜  ~ ls -l ${ZSH_CACHE_DIR}/kubectl_completion
-rw-r--r-- 1 eric users 459559 Mar  6  2021 /home/eric/.oh-my-zsh/cache/kubectl_completion
➜  ~ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.3", GitCommit:"c92036820499fedefec0f847e2054d824aea6cd1", GitTreeState:"archive", BuildDate:"2021-10-28T06:55:39Z", GoVersion:"go1.17.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:32:41Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
➜  ~ for plugin ($plugins); do
  timer=$(($(date +%s%N)/1000000))
  if [ -f $ZSH_CUSTOM/plugins/$plugin/$plugin.plugin.zsh ]; then
    source $ZSH_CUSTOM/plugins/$plugin/$plugin.plugin.zsh
  elif [ -f $ZSH/plugins/$plugin/$plugin.plugin.zsh ]; then
    source $ZSH/plugins/$plugin/$plugin.plugin.zsh
  fi
  now=$(($(date +%s%N)/1000000))
  elapsed=$(($now-$timer))
  echo $elapsed ms":" $plugin
done
37 ms: git
25 ms: fzf
3 ms: tmux
4 ms: kubectl
1 ms: docker
➜  ~
```

**Comment by carlosala** [comment]
This was fixed in `658eb01d`.
